Refactor implicit Self param into a member on SemIR::Function

dwblaikie opened the pull request:
This ensures the data is available for more uses (specifically for diagnosing virtual/abstract/impl functions on non-instance methods).

It still doesn't quite address the TODO to move the Self param search all the way back to the param walk in all cases. To do that in the case that still has a separate search loop, I think we'd have to change `Check::NameComponent` to carry this information (as it carries the implicit_param_patterns-id) - though there's comments in NameComponent suggesting it shouldn't carry function-specific things like `call_params_id` and `return_slot_pattern_id` - so I wasn't sure if it was suitable to add more there, but I can - possibly in a follow-up change.

dwblaikie committed:
Refactor implicit Self param into a member on SemIR::Function

This ensures the data is available for more uses (specifically for
diagnosing virtual/abstract/impl functions on non-instance methods).

It still doesn't quite address the TODO to move the Self param search
all the way back to the param walk in all cases. To do that in the case
that still has a separate search loop, I think we'd have to change
`Check::NameComponent` to carry this information (as it carries the
implicit_param_patterns-id) - though there's comments in NameComponent
suggesting it shouldn't carry function-specific things like
`call_params_id` and `return_slot_pattern_id` - so I wasn't sure if it
was suitable to add more there, but I can - possibly in a follow-up
change.

danakj commented on toolchain/sem_ir/function.h:
```diff
@@ -31,6 +31,8 @@ struct FunctionFields {
   // this function.
   VirtualModifier virtual_modifier;
 
+  InstId self_param_id = SemIR::InstId::None;
```
Document?

dwblaikie replied:
cf330b3ea16f91922aca27bc24178edf779647d8

danakj commented on toolchain/check/import_ref.cpp:
```diff
@@ -977,6 +977,9 @@ static auto LoadLocalPatternConstantIds(ImportRefResolver& resolver,
 // LoadLocalPatternConstantIds(param_patterns_id) has completed without adding
 // any new work to work_stack_.
 //
+// `self_param_id` is an optional out parameter, populated with the InstId in
```
The TODO below is about the paragraph above, I think? So maybe we should move this below the TODO paragraph?

dwblaikie replied:
8a5d3e46040f023cfa0ed48b1848d26886fd680e

danakj commented on toolchain/check/handle_function.cpp:
```diff
@@ -263,6 +264,17 @@ static auto BuildFunctionDecl(Context& context,
+    if (SemIR::Function::GetNameFromPatternId(
+            context.sem_ir(), implicit_param_id) == SemIR::NameId::SelfValue) {
+      CARBON_CHECK(!function_info.self_param_id.has_value());
+      function_info.self_param_id = implicit_param_id;
```
nit: This could set a local variable before the Function construction, and then we could unconditionaly set .self_param_id during construction, in the aggregate initializer, instead of mutating `function_info` afterward?

dwblaikie replied:
0d33379db7fc8a4bc74960e04a9e2413fbce1ac0

danakj reviewed (approved):
Drive-by LGTM with a couple comments

jonmeow committed:
Fold Context::GetCurrentScopeAs back into ScopeStack (#4936)

This adds a `SemIR::File` pointer to `scope_stack` so that
`GetCurrentScopeAs` doesn't require the argument, which would be why the
helper existed on `context`.

dwblaikie committed:
Compute self_param_id before constructing the SemIR::Function

Also switch to use llvm::find_if rather than an explicit loop/if.

dwblaikie committed:
Move param comment to avoid splitting a TODO from its context

dwblaikie committed:
Document the new SemIR::Function::self_param_id member

dwblaikie commented:
> Drive-by LGTM with a couple comments

Thanks!

chandlerc commented:
(Just to be clear, glad you're not waiting on me, and thanks Dana for handling this review!)

dwblaikie committed:
Merge branch 'trunk' into virtual_must_be_instance_method

danakj commented:
Looks like tidy is unhappy

dwblaikie committed:
Add explicit const and * to auto.

dwblaikie commented:
> Looks like tidy is unhappy

Hmm, yeah - it looked like it was getting wedged/taking hours yesterday, but now I see an actual diagnostic (though not one I see locally - version skew, I guess). Made a fix/update, will see how it goes

danakj commented:
Hm I left a comment.. somewhere.. not showing in the UI here so I will repost it:

IIUC this `i` is an iterator. Using `const auto*` here is both incorrect and a source regression. The iterator type should be abstract, and allowed to change from a pointer to a bounds-checked type, etc, freely.

Chromium has a plugin that enforces `auto*` for pointers, but it explicitly does not look through type aliases to avoid this issue.

If I understood wrong, then all good. But if this is in fact an iterator, we should revert this change and disable the tidy check.
